src/connection.rs

```diff
@@ -0,0 +1,205 @@
+use byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
+use openssl::ssl::*;
+use protobuf::{CodedOutputStream, Message};
+use std::io::{Cursor, Read};
+use std::net::TcpStream;
+use std::path::Path;
+use std::sync::Arc;
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+
+use message_types::MessageType::*;
+use message_types::*;
+use mumble;
+
+// Connection timeout in seconds. It waits this long if no message was received
+const TIMEOUT: u64 = 5;
+
+// How long to wait between pings
+const PING_DELAY: u64 = 10;
+
+pub struct Connection {
+    stream: SslStream<TcpStream>,
+    session: Option<u32>,
+    users: Vec<mumble::UserState>,
+    // TODO: channels: Vec<mumble::ChannelState>,
+    last_ping: u64,
+}
+
+impl Connection {
+    pub fn new(url: &str, key: &Path, cert: &Path) -> Connection {
+        // Setup TCP stream
+        let ctx = Connection::setup_tls(key, cert);
+        let tcp = TcpStream::connect(url).unwrap();
+        tcp.set_read_timeout(Some(Duration::new(TIMEOUT, 0)))
+            .unwrap();
+        let stream = Ssl::new(&ctx).unwrap().connect(tcp).unwrap();
+
+        Self {
+            stream,
+            session: None,
+            users: vec![],
+            //channels: vec![],
+            last_ping: 0,
+        }
+    }
+
+    fn setup_tls(key: &Path, cert: &Path) -> Arc<SslContext> {
+        let mut ctx = SslContext::builder(SslMethod::tls()).unwrap();
+        ctx.set_verify(SslVerifyMode::NONE);
+        ctx.set_certificate_chain_file(cert).unwrap();
+        ctx.set_private_key_file(&key, SslFiletype::PEM).unwrap();
+        ctx.set_cipher_list("ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:ECDHE-ECDSA-DES-CBC3-SHA:ECDHE-RSA-DES-CBC3-SHA:DES-CBC3-SHA:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!aECDH:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SHA:!KRB5-DES-CBC3-SHA").unwrap();
```

Comment:
Should this be hardcoded and non-overridable? Does Mumble enforce this set of ciphers in it's standard?